Implementace CacheDecorator

JindrichPilar opened the pull request:
Cache pozadavku na server.

JindrichPilar committed:
Pridano WebServiceInterface
Misto magicke metody pridana metoda call

JindrichPilar committed:
Implementace CacheDecorator resolves #44
Pridano CacheInterface a vzorova implementace ArrayCache

xificurk commented on src/Wsdl/Decorator/Cache/CacheDecorator.php:
```diff
@@ -0,0 +1,46 @@
+        $callHash = $this->hashCall($functionName, $arguments);
+
+	if ($this->cache->has($callHash)) {
+            return $this->cache->get($callHash);
```
Tady si zaděláváme na race condition, pokud bude uložení keše obstarávat nějaká služba dostupná pro více vláken.
Mělo by se rovnou číst a pokud je něco uloženo (tj. ne `NULL`), tak vrátit, jinak pokračovat dotazem a uložením do keš.

JindrichPilar replied:
Opraveno

xificurk commented:
A ještě bych byl pro to, aby se zachovala možnost používat WS pomocí magického `__call()`.

JindrichPilar committed:
CacheInterface metoda has odstranena

JindrichPilar committed:
CacheDecorator vyzaduje alespon 1 uspesny request

JindrichPilar commented:
Pozadavek na 1 uspesny request na server pridan. Pripraveno k merge.

@xificurk  Mas nejaky specificky duvod proc `__call()` zachovat? Je pravda ze s tim jsou pozadavky na server citelnejsi. Ale osobne mi prijde ze `__call()` znehodnocuje kodovani vuci interface. (Pouze osobni pocit)

xificurk commented:
@JindrichPilar Je to čitelnější + aspoň nějaká zpětná kompatibilita.

JindrichPilar committed:
WebService __call method vracena
 - ``__call`` vyzadovana z ``WebServiceInterface``
 - ``__call`` implementovana v ``AbstractDecorator`` jako alias metody ``call``

JindrichPilar commented:
Metoda `__call` doplnena. Myslim ze je to kompletni a pripraveno k merge.
